File: cinema-frontend/src/app/models/movie.ts
export interface Movie {
  id?: number;
  title: string;
  originalTitle?: string; // English title
  description: string;
  duration: number;
  releaseDate: Date;
  genre: string;
  director: string;
  cast: string;
  posterUrl?: string;
  trailerUrl?: string;
  rating: number;
  status: 'NOW_SHOWING' | 'COMING_SOON' | 'ENDED';
  filmRating?: string;
  price: number;
  episodes?: number; // Number of episodes for series
  createdAt?: Date;
  updatedAt?: Date;
}

export interface MovieDTO {
  title: string;
  originalTitle?: string; // English title
  description: string;
  duration: number;
  releaseDate: Date;
  genre: string;
  director: string;
  cast: string;
  rating: number;
  status: string;
  filmRating?: string;
  price: number;
  episodes?: number; // Number of episodes for series
  posterFile?: File;
  bannerFile?: File;
  [key: string]: any; // Allow dynamic property access
}

File: src/main/java/com/project/cinema/movie/Models/Movie.java
package com.project.cinema.movie.Models;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.*;
import org.springframework.data.annotation.CreatedDate;

import java.util.Date;
import java.util.List;

@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
@Entity
@Data
@Table(name = "Movies")
public class Movie {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, unique = true, columnDefinition = "TEXT CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci")
    private String title;

    @Column(columnDefinition = "LONGTEXT CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci")
    private String description;
    private int duration;

    @Temporal(TemporalType.DATE)
    @JsonFormat(pattern = "yyyy-MM-dd")
    private Date releaseDate;

    @Column(columnDefinition = "TEXT CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci")
    private String genre;
    @Column(columnDefinition = "TEXT CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci")
    private String director;
    private String posterUrl;
    private String trailerUrl;
    @Column(columnDefinition = "TEXT CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci")
    private String language;
    @Column(columnDefinition = "TEXT CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci")
    private String cast;
    private double rating;
    private String status;
    private double price;
    private FilmRating filmRating;
    @CreatedDate
    @Temporal(TemporalType.TIMESTAMP)
    private Date createdAt;
    @Temporal(TemporalType.TIMESTAMP)
    private Date updatedAt;
    @JsonIgnore
    @OneToMany(mappedBy = "movie", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private List<Showtime> showtimes; // Broadcast schedule for the movie
    @PrePersist
    protected void onCreate() {
        this.createdAt = new Date(); // Thiết lập createdAt với ngày giờ hiện tại
    }

    public Movie(String title, String description, int duration, Date releaseDate, String genre, String director, String posterUrl, String trailerUrl, String language, String cast, double rating, String status, double price, FilmRating filmRating) {
        this.title = title;
        this.description = description;
        this.duration = duration;
        this.releaseDate = releaseDate;
        this.genre = genre;
        this.director = director;
        this.posterUrl = posterUrl;
        this.trailerUrl = trailerUrl;
        this.language = language;
        this.cast = cast;
        this.rating = rating;
        this.status = status;
        this.price = price;
        this.filmRating = filmRating;
    }
}

